### PR TITLE
[llvm-objdump][AMDGPU] Pass ELF ABIVersion through disassembler

### PR DESCRIPTION
**Pull request by epilk**
Admittedly, its a bit ugly to pass the ABIVersion through onSymbolStart but I'm not sure what a better place for it would be.

### CODE REVIEW

**Comment by MaskRay** [comment]
> Admittedly, its a bit ugly to pass the ABIVersion through onSymbolStart but I'm not sure what a better place for it would be.

I am indeed concerned about this. Can you place the version inside `AMDGPUDisassembler` instead?

**Comment by jh7370** [comment]
> > Admittedly, its a bit ugly to pass the ABIVersion through onSymbolStart but I'm not sure what a better place for it would be.
> 
> I am indeed concerned about this. Can you place the version inside `AMDGPUDisassembler` instead?

+1. The ABIVersion should be constant for a given file, so it feels like it should be configured up front when configuring the disassembler (and therefore presumably stored somewhere inside the disassembler class).

**Comment by epilk** [comment]
> > > Admittedly, its a bit ugly to pass the ABIVersion through onSymbolStart but I'm not sure what a better place for it would be.
> > 
> > 
> > I am indeed concerned about this. Can you place the version inside `AMDGPUDisassembler` instead?
> 
> +1. The ABIVersion should be constant for a given file, so it feels like it should be configured up front when configuring the disassembler (and therefore presumably stored somewhere inside the disassembler class).

Okay, I added a `virtual setABIVersion()` function to `MCDisassembler` in the update, so we can just store it in the disassembler. WDYT?

File: llvm/lib/Target/AMDGPU/Disassembler/AMDGPUDisassembler.cpp [hunk 1]
```diff
@@ -2184,8 +2188,7 @@ AMDGPUDisassembler::decodeKernelDescriptorDirective(
                       KERNEL_CODE_PROPERTY_ENABLE_WAVEFRONT_SIZE32);
     }
 
-    // FIXME: We should be looking at the ELF header ABI version for this.
-    if (AMDGPU::getDefaultAMDHSACodeObjectVersion() >= AMDGPU::AMDHSA_COV5)
+    if (CodeObjectVersion >= AMDGPU::AMDHSA_COV5)
```
Comment (kzhuravl) [inline llvm/lib/Target/AMDGPU/Disassembler/AMDGPUDisassembler.cpp:2191]: What is the difference between AMDHSA_COV5 and ELFABIVERSION_AMDGPU_HSA_V5? Can't we use ELFABIVERSION_AMDGPU_HSA_V5 here?

Comment (epilk) [reply llvm/lib/Target/AMDGPU/Disassembler/AMDGPUDisassembler.cpp:2191]: `ELFABIVERSION_AMDGPU_HSA_V5` is the value encoded in the ELF header, which is 3 for COV5 and 2 for COV4. `AMDHSA_COV5` is the actual version, so it has value 5. `getELFABIVersion()` maps from actual to encoded version, and `getAMDHSACodeObjectVersion()` maps from encoded to actual version. It's unfortunate that there's a difference :/

File: llvm/lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.h [hunk 2]
```diff
@@ -50,6 +50,9 @@ bool isHsaAbi(const MCSubtargetInfo &STI);
 /// \returns Code object version from the IR module flag.
 unsigned getAMDHSACodeObjectVersion(const Module &M);
 
+/// \returns Code object version from ELF's e_ident[ABIVERSION].
```
Comment (MaskRay) [inline llvm/lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.h:53]: `e_ident[EI_ABIVERSION]`

File: llvm/test/tools/llvm-objdump/ELF/AMDGPU/kd-cov5.s [hunk 3]
```diff
@@ -0,0 +1,23 @@
+; RUN: sed 's/CODE_OBJECT_VERSION/5/g' %s \
+; RUN:   | llvm-mc --triple=amdgcn-amd-amdhsa -mcpu=gfx1010 -mattr=-xnack,+wavefrontsize32,-wavefrontsize64 -filetype=obj > %t.o
+; RUN: llvm-objdump --disassemble-symbols=kernel.kd %t.o | FileCheck %s --check-prefixes=COV5,CHECK
+
+; RUN: sed 's/CODE_OBJECT_VERSION/4/g' %s \
+; RUN:   | llvm-mc --triple=amdgcn-amd-amdhsa -mcpu=gfx1010 -mattr=-xnack,+wavefrontsize32,-wavefrontsize64 -filetype=obj > %t.o
+; RUN: llvm-objdump --disassemble-symbols=kernel.kd %t.o | FileCheck %s --check-prefixes=COV4,CHECK
+
+; Verify that .amdhsa_uses_dynamic_stack is only printed on COV5+.
```
Comment (MaskRay) [inline llvm/test/tools/llvm-objdump/ELF/AMDGPU/kd-cov5.s:9]: `;; ` for a non-RUN-non-CHECK comment

**Review by MaskRay — APPROVED** [review]

**Review by jh7370 — COMMENTED** [review]
ELFObjectFile and llvm-obdjump parts LGTM, I don't know anything really to be able to review the other aspects.

**Review by kzhuravl — APPROVED** [review]
AMDGPU-related parts LGTM, thanks!

**Pull request merged**